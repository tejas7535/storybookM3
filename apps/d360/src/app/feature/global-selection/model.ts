import { GlobalSelectionStateService } from '../../shared/components/global-selection-criteria/global-selection-state.service';
import { SelectableValue } from '../../shared/components/inputs/autocomplete/selectable-values.utils';

export interface CustomerEntry {
  salesOrg?: string;
  customerNumber: string;
  customerName?: string;
  planningCurrency?: string;
}

export type GlobalSelectionCriteriaFilters = Partial<
  Record<(typeof GlobalSelectionStateService.stateKeys)[number], string[]>
>;

export type GlobalSelectionCriteriaFields = Partial<
  Record<
    (typeof GlobalSelectionStateService.stateKeys)[number],
    SelectableValue[]
  >
>;

export enum GlobalSelectionStatus {
  DATA_AVAILABLE = 'DATA_AVAILABLE',
  DATA_NO_RESULTS = 'DATA_NO_RESULTS',
  DATA_LOADING = 'DATA_LOADING',
  DATA_ERROR = 'DATA_ERROR',
  DATA_NOTHING_SELECTED = 'DATA_NOTHING_SELECTED',
}

export enum Region {
  Europe = 'EU',
  GreaterChina = 'GC',
  AsiaPacific = 'AP',
  Americas = 'AM',
}
