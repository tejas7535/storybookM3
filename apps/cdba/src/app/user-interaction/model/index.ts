export * from './action.model';
export * from './interaction.model';
export * from './interaction-type.enum';
