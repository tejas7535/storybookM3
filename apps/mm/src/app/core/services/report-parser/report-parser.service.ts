/* eslint-disable max-lines */
import { inject, Injectable } from '@angular/core';

import { CalculationResultReportInput } from '@mm/core/store/models/calculation-result-report-input.model';
import {
  CalculationResult,
  MountingTools,
  PumpItem,
  ReportMessages,
  ResultItem,
  ResultItemWithTitle,
} from '@mm/core/store/models/calculation-result-state.model';

import {
  BLOCK,
  STRING_ERROR_BLOCK,
  STRING_NOTE_BLOCK,
  STRING_OUTP_ADDITIONAL_TOOLS,
  STRING_OUTP_CHECK_VALUES_FOR_CLEARANCE_CLASSES,
  STRING_OUTP_END_POSITION,
  STRING_OUTP_HYDRAULIC_NUT,
  STRING_OUTP_INPUT,
  STRING_OUTP_LOCKNUT,
  STRING_OUTP_MOUNTING_RECOMMENDATIONS_IDMM_BEARING_SEAT_IDMM_MOUNTING_METHOD,
  STRING_OUTP_MOUNTING_TOOLS_AND_UTILITIES,
  STRING_OUTP_PUMPS,
  STRING_OUTP_RADIAL_CLEARANCE_REDUCTION_AND_AXIAL_DISPLACEMENT,
  STRING_OUTP_RESULTS,
  STRING_OUTP_SLEEVE_CONNECTORS,
  STRING_OUTP_STARTING_POSITION,
  STRING_WARNING_BLOCK,
  TABLE,
  TEXT,
  VARIABLE_BLOCK,
  VARIABLE_LINE,
} from '../bearinx-result.constant';
import {
  BearinxOnlineResult,
  BearinxOnlineResultSubordinate,
} from '../bearinx-result.interface';
import { ResultPositionsPriorityService } from '../result-positions-priority/result-positions-priority.service';

@Injectable({
  providedIn: 'root',
})
export class ReportParserService {
  private readonly resultPositionsPriorityService = inject(
    ResultPositionsPriorityService
  );

  public parseResponse(response: BearinxOnlineResult): CalculationResult {
    const startPositions = this.extractStartPositions(response);
    const endPositions = this.extractEndPositions(response);

    return {
      startPositions,
      endPositions,
      mountingRecommendations: this.extractMountingRecommendations(response),
      mountingTools: this.extractMountingTools(response),
      radialClearance:
        this.extractRadialClearanceReductionAndAxialDisplacement(response),
      clearanceClasses: this.extractClearanceClasses(response),
      reportMessages: this.extractReportMessages(response),
      inputs: this.parseInputResponse(response),
    };
  }

  private extractStartPositions(response: BearinxOnlineResult): ResultItem[] {
    const unorderedStartPositions = this.extractItemsFromPath(response, [
      { titleID: STRING_OUTP_RESULTS, identifier: BLOCK },
      { titleID: STRING_OUTP_STARTING_POSITION, identifier: VARIABLE_BLOCK },
    ]);

    return this.resultPositionsPriorityService.getPrioritizedStartItems(
      unorderedStartPositions
    );
  }

  private extractEndPositions(response: BearinxOnlineResult): ResultItem[] {
    const unorderedEndPositions = this.extractItemsFromPath(response, [
      { titleID: STRING_OUTP_RESULTS, identifier: BLOCK },
      { titleID: STRING_OUTP_END_POSITION, identifier: VARIABLE_BLOCK },
    ]);

    return this.resultPositionsPriorityService.getPrioritizedEndItems(
      unorderedEndPositions
    );
  }

  private extractMountingTools(response: BearinxOnlineResult): MountingTools {
    return {
      additionalTools: this.extractAdditionalTools(response),
      hydraulicNut: this.extractHydraulicNut(response),
      pumps: this.extractPumps(response),
      locknut: this.extractLockNut(response),
      sleeveConnectors: this.extractSleeveConnectors(response),
    };
  }

  private extractReportMessages(response: BearinxOnlineResult): ReportMessages {
    return {
      errors: this.extractMessageByTitleId(response, STRING_ERROR_BLOCK),
      notes: this.extractMessageByTitleId(response, STRING_NOTE_BLOCK),
      warnings: this.extractMessageByTitleId(response, STRING_WARNING_BLOCK),
    };
  }

  private extractMessageByTitleId(
    originalResult: BearinxOnlineResult,
    titleId: string
  ): string[] {
    const msg = this.extractSubordinatesByTitleID(titleId, originalResult);

    return this.formatMessageSubordinates(msg);
  }

  private extractSubordinatesByTitleID(
    titleId: string,
    input: BearinxOnlineResult
  ): BearinxOnlineResultSubordinate[] {
    return input.subordinates.filter(
      (subordinate) => subordinate.titleID === titleId
    );
  }

  private extractAdditionalTools(
    originalResult: BearinxOnlineResult
  ): ResultItem[] {
    return this.extractItemsFromPath(
      originalResult,
      this.getMountingToolsItemPath(
        STRING_OUTP_ADDITIONAL_TOOLS,
        VARIABLE_BLOCK
      )
    );
  }

  private extractHydraulicNut(
    originalResult: BearinxOnlineResult
  ): ResultItem[] {
    return this.extractItemsFromPath(
      originalResult,
      this.getMountingToolsItemPath(STRING_OUTP_HYDRAULIC_NUT, VARIABLE_BLOCK)
    );
  }

  private extractLockNut(originalResult: BearinxOnlineResult): ResultItem[] {
    return this.extractItemsFromPath(
      originalResult,
      this.getMountingToolsItemPath(STRING_OUTP_LOCKNUT, VARIABLE_BLOCK)
    );
  }

  private extractSleeveConnectors(
    originalResult: BearinxOnlineResult
  ): ResultItem[] {
    return this.extractItemsFromPath(
      originalResult,
      this.getMountingToolsItemPath(
        STRING_OUTP_SLEEVE_CONNECTORS,
        VARIABLE_BLOCK
      )
    );
  }

  private getMountingToolsItemPath(titleId: string, identifier: string) {
    return [
      { titleID: STRING_OUTP_RESULTS, identifier: BLOCK },
      {
        titleID: STRING_OUTP_MOUNTING_TOOLS_AND_UTILITIES,
        identifier: BLOCK,
      },
      { titleID: titleId, identifier },
    ];
  }

  private extractRadialClearanceReductionAndAxialDisplacement(
    originalResult: BearinxOnlineResult
  ): ResultItem[] {
    const radialClearanceResultData = this.extractTableItemsWithTitle(
      originalResult,
      STRING_OUTP_RADIAL_CLEARANCE_REDUCTION_AND_AXIAL_DISPLACEMENT
    );

    return this.resultPositionsPriorityService.getPrioritizedAndFormattedRadialClearance(
      radialClearanceResultData
    );
  }

  private extractClearanceClasses(
    originalResult: BearinxOnlineResult
  ): ResultItem[] {
    const clearanceClassesResultData = this.extractTableItemsWithTitle(
      originalResult,
      STRING_OUTP_CHECK_VALUES_FOR_CLEARANCE_CLASSES
    );
    const clearanceClasses =
      this.resultPositionsPriorityService.getPrioritizedClearanceClasses(
        clearanceClassesResultData
      );

    return clearanceClasses;
  }

  private extractTableItemsWithTitle(
    originalResult: BearinxOnlineResult,
    titleId: string
  ): ResultItemWithTitle[] {
    const subordinates: BearinxOnlineResultSubordinate =
      this.extractSubordinatesFromPath(originalResult, [
        { titleID: STRING_OUTP_RESULTS, identifier: BLOCK },
        { titleID: titleId, identifier: TABLE },
      ]);

    return (
      subordinates?.data?.items.map((item) => {
        const title = item[0].value;
        const items = item.slice(1).map((fieldItem) => ({
          designation: fieldItem.field,
          value: fieldItem.value,
          unit: fieldItem.unit,
        }));

        return {
          title,
          items,
        };
      }) || []
    );
  }

  private extractPumps(originalResult: BearinxOnlineResult): {
    title: string;
    items: PumpItem[];
  } {
    const pumpsSubordinates: BearinxOnlineResultSubordinate =
      this.extractSubordinatesFromPath(
        originalResult,
        this.getMountingToolsItemPath(STRING_OUTP_PUMPS, TABLE)
      );

    const items =
      pumpsSubordinates?.data?.items
        .filter(([firstItem, secondItem]) => firstItem && secondItem)
        .map(([firstItem, secondItem]) => ({
          isRecommended: firstItem.value === 'recommended',
          field: secondItem.field,
          value: secondItem.value,
        })) || [];

    return {
      title: this.extractPumpsTitle(pumpsSubordinates),
      items,
    };
  }

  private extractPumpsTitle(
    pumpsSubordinates: BearinxOnlineResultSubordinate
  ): string {
    const fields = pumpsSubordinates?.data?.fields || [];

    switch (fields.length) {
      case 2:
      case 3:
        return fields[1];
      case 4:
        return fields[1] + fields[2];
      default:
        return '';
    }
  }

  private extractMountingRecommendations(
    originalResult: BearinxOnlineResult
  ): string[] {
    const mountingRecommendationsValue: string[] = [];
    const path = [
      { titleID: STRING_OUTP_RESULTS, identifier: BLOCK },
      {
        titleID:
          STRING_OUTP_MOUNTING_RECOMMENDATIONS_IDMM_BEARING_SEAT_IDMM_MOUNTING_METHOD,
        identifier: BLOCK,
      },
    ];

    const mountingRecommendations = this.extractSubordinatesFromPath(
      originalResult,
      path
    );

    if (!mountingRecommendations) {
      return mountingRecommendationsValue;
    }

    const mountingRecommendationsWithDashes = this.formatMessageSubordinates(
      mountingRecommendations.subordinates
    );

    return this.removeLeadingDashFromMountingRecommendations(
      mountingRecommendationsWithDashes
    );
  }

  private removeLeadingDashFromMountingRecommendations(
    originalMountingRecommendations: string[]
  ): string[] {
    return originalMountingRecommendations.map((recommendation) =>
      recommendation.replace(/^-/, '').trim()
    );
  }

  private extractSubordinatesFromPath(
    input: BearinxOnlineResult | BearinxOnlineResultSubordinate,
    path: Partial<BearinxOnlineResultSubordinate>[]
  ): BearinxOnlineResult | BearinxOnlineResultSubordinate | undefined {
    let result: BearinxOnlineResult | BearinxOnlineResultSubordinate = input;

    for (const pathItem of path) {
      // find sub item by identifier and designation
      const item = result.subordinates.find((subordinate) =>
        this.matchItem(subordinate, pathItem)
      );

      if (!item) {
        return undefined;
      }

      result = item;
    }

    return result;
  }

  private matchItem(
    subordinate: BearinxOnlineResultSubordinate,
    objectToMatch: Partial<BearinxOnlineResultSubordinate>
  ): boolean {
    return Object.entries(objectToMatch).every(([key, value]) =>
      key in subordinate
        ? subordinate[key as keyof BearinxOnlineResultSubordinate] === value
        : value === undefined
    );
  }

  private formatMessageSubordinates(
    rawSubordinates: BearinxOnlineResultSubordinate[]
  ): string[] {
    return rawSubordinates
      .flatMap((subordinate) => {
        if (subordinate.identifier === TEXT) {
          return subordinate.text;
        }
        if (subordinate.identifier === BLOCK) {
          return this.formatMessageSubordinates(subordinate.subordinates);
        }

        return [];
      })
      .filter((e) => String(e).trim());
  }

  private parseInputResponse(
    response: BearinxOnlineResult
  ): CalculationResultReportInput[] {
    const path = [{ titleID: STRING_OUTP_INPUT, identifier: BLOCK }];

    const inputData = this.extractSubordinatesFromPath(response, path);

    if (!inputData?.subordinates) {
      return [];
    }

    return this.formatReportInputResult(inputData.subordinates) ?? [];
  }

  private formatReportInputResult(
    input: BearinxOnlineResultSubordinate[]
  ): CalculationResultReportInput[] {
    return input?.map((reportInput) => this.getReportInput(reportInput)) ?? [];
  }

  private getReportInput(
    input: BearinxOnlineResultSubordinate
  ): CalculationResultReportInput {
    const result: CalculationResultReportInput = {
      hasNestedStructure:
        input.identifier === BLOCK || input.identifier === TABLE,
      title: input?.title,
      titleID: input?.titleID,
    };

    switch (input.identifier) {
      case BLOCK:
      case VARIABLE_BLOCK:
        result.subItems = this.formatReportInputResult(input.subordinates);
        break;
      case TABLE:
        result.subItems = this.extractReportInputFromTableSubordinate(input);
        break;
      case VARIABLE_LINE:
        result.designation = input?.designation;
        result.abbreviation = input?.abbreviation;
        result.value = input?.value;
        result.unit = input?.unit;
        break;
      default:
        throw new Error(`Unexpected identifier: ${input.identifier}`);
    }

    return result;
  }

  private extractReportInputFromTableSubordinate(
    input: Partial<BearinxOnlineResultSubordinate>
  ): CalculationResultReportInput[] {
    if (!input?.data) {
      return [];
    }

    const { items } = input.data;

    if (items.length > 1) {
      throw new Error('Only one line tables supported');
    }

    return items[0].map((tableRow) => ({
      hasNestedStructure: false,
      designation: tableRow.field,
      value: tableRow.value,
      unit: tableRow?.unit,
    }));
  }

  private extractItemsFromPath(
    originalResult: BearinxOnlineResult,
    path: { titleID: string; identifier: string }[]
  ): ResultItem[] {
    const resultBlock = this.extractSubordinatesFromPath(originalResult, path);

    if (!resultBlock?.subordinates) {
      return [];
    }

    return resultBlock.subordinates.map(
      (subordinate: BearinxOnlineResultSubordinate) => ({
        abbreviation: subordinate.abbreviation,
        designation: subordinate.designation,
        unit: subordinate.unit,
        value: subordinate.value,
      })
    );
  }
}
