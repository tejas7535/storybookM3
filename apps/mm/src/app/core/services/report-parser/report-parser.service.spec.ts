import { CalculationResult } from '@mm/core/store/models/calculation-result-state.model';
import { createServiceFactory, SpectatorService } from '@ngneat/spectator/jest';

import { BLOCK, STRING_OUTP_RESULTS } from '../bearinx-result.constant';
import { BearinxOnlineResult } from '../bearinx-result.interface';
import {
  JSON_REPORT_ALL_TYPES_INPUTS_MOCK,
  JSON_REPORT_FULL_RESPONSE_MOCK,
  JSON_REPORT_INCORRECT_INPUT_MOCK,
  JSON_REPORT_LOCKNUT_AND_CHECK_VALUES_MOCK,
  JSON_REPORT_THERMAL_BEARING_MOCK,
} from './../../../../testing/mocks/report-parser.mock';
import { ReportParserService } from './report-parser.service';

describe('ReportParserService', () => {
  let spectator: SpectatorService<ReportParserService>;
  let service: ReportParserService;
  let result: CalculationResult;
  const createService = createServiceFactory(ReportParserService);

  beforeEach(() => {
    spectator = createService();
    service = spectator.service;
  });

  it('should be created', () => {
    expect(service).toBeTruthy();
  });

  describe('isRecommendedPump', () => {
    let isRecommendedPump: (value: string) => boolean;

    beforeEach(() => {
      isRecommendedPump = (service as any)['isRecommendedPump'].bind(service);
    });

    it('should identify "recommended" as a recommended pump', () => {
      expect(isRecommendedPump('recommended')).toBe(true);
    });

    it('should identify "empfohlen" (German) as a recommended pump', () => {
      expect(isRecommendedPump('empfohlen')).toBe(true);
    });

    it('should identify "se recomienda" (Spanish) as a recommended pump', () => {
      expect(isRecommendedPump('se recomienda')).toBe(true);
    });

    it('should identify "recommandé" (French) as a recommended pump', () => {
      expect(isRecommendedPump('recommandé')).toBe(true);
    });

    it('should identify "рекомендуется" (Russian) as a recommended pump', () => {
      expect(isRecommendedPump('рекомендуется')).toBe(true);
    });

    it('should identify "推荐" (Chinese) as a recommended pump', () => {
      expect(isRecommendedPump('推荐')).toBe(true);
    });

    it('should not identify "alternative" as a recommended pump', () => {
      expect(isRecommendedPump('alternative')).toBe(false);
    });

    it('should not identify empty string as a recommended pump', () => {
      expect(isRecommendedPump('')).toBe(false);
    });

    it('should not identify any other string as a recommended pump', () => {
      expect(isRecommendedPump('some other value')).toBe(false);
    });
  });

  describe('parseResponse', () => {
    describe('when result is defined', () => {
      beforeEach(() => {
        result = service.parseResponse(JSON_REPORT_FULL_RESPONSE_MOCK);
      });

      it('should parse response end positions data', () => {
        expect(result.endPositions).toMatchSnapshot();
      });

      it('should parse mounting recommendations data', () => {
        expect(result.mountingRecommendations).toMatchSnapshot();
      });

      it('should have empty errors', () => {
        expect(result.reportMessages.errors).toMatchSnapshot();
      });

      it('should have warnings', () => {
        expect(result.reportMessages.warnings).toMatchSnapshot();
      });

      it('should have notes', () => {
        expect(result.reportMessages.notes).toMatchSnapshot();
      });

      it('should have inputs', () => {
        expect(result.inputs).toMatchSnapshot();
      });

      it('should have pumps with recommended and alternative boolean flag', () => {
        expect(result.mountingTools.pumps).toMatchSnapshot();
      });

      it('should have nut data', () => {
        expect(result.mountingTools.hydraulicNut).toMatchSnapshot();
      });
    });

    describe('when result is undefined', () => {
      beforeEach(() => {
        result = service.parseResponse({
          subordinates: [
            {
              titleID: STRING_OUTP_RESULTS,
              identifier: BLOCK,
              subordinates: [],
            },
          ],
        } as Partial<BearinxOnlineResult> as BearinxOnlineResult);
      });

      it('should return empty end positions', () => {
        expect(result.endPositions).toEqual([]);
      });

      it('should return empty mounting recommendations', () => {
        expect(result.mountingRecommendations).toEqual([]);
      });

      it('should have empty errors', () => {
        expect(result.reportMessages.errors).toEqual([]);
      });

      it('should have empty warnings', () => {
        expect(result.reportMessages.warnings).toEqual([]);
      });

      it('should have empty notes', () => {
        expect(result.reportMessages.notes).toEqual([]);
      });

      it('should have empty pumps', () => {
        expect(result.mountingTools.pumps).toEqual({ title: '', items: [] });
      });

      it('should have empty nut data', () => {
        expect(result.mountingTools.hydraulicNut).toEqual([]);
      });

      it('should have empty radial clearance data', () => {
        expect(result.radialClearance).toEqual([]);
      });

      it('should have empty clearance classes data', () => {
        expect(result.clearanceClasses).toEqual([]);
      });

      it('should have empty sleeve connectors data', () => {
        expect(result.mountingTools.sleeveConnectors).toEqual([]);
      });

      it('should have empty input data', () => {
        expect(result.inputs).toEqual([]);
      });
    });
  });

  describe('parse Response with checknut and clearances', () => {
    describe('when result is defined', () => {
      beforeEach(() => {
        result = service.parseResponse(
          JSON_REPORT_LOCKNUT_AND_CHECK_VALUES_MOCK
        );
      });

      it('should parse radial clearance data', () => {
        expect(result.radialClearance).toMatchSnapshot();
      });

      it('should parse clearance classes data', () => {
        expect(result.clearanceClasses).toMatchSnapshot();
      });

      it('should parse locknut data', () => {
        expect(result.mountingTools.locknut).toMatchSnapshot();
      });

      it('should parse sleeve connectors data', () => {
        expect(result.mountingTools.sleeveConnectors).toMatchSnapshot();
      });

      it('should have empty notes section', () => {
        expect(result.reportMessages.notes).toMatchSnapshot();
      });
    });
  });

  describe('when parsing incorrect input response', () => {
    it('should throw error', () => {
      expect(() =>
        service.parseResponse(
          JSON_REPORT_INCORRECT_INPUT_MOCK as BearinxOnlineResult
        )
      ).toThrow(new Error('Unexpected identifier: incorrect identifier'));
    });
  });

  describe('when parsing input with all possible identifier types', () => {
    beforeEach(() => {
      result = service.parseResponse(
        JSON_REPORT_ALL_TYPES_INPUTS_MOCK as BearinxOnlineResult
      );
    });

    it('should parse inputs', () => {
      expect(result.inputs).toMatchSnapshot();
    });
  });

  describe('when parsing thermal bearing response', () => {
    beforeEach(() => {
      result = service.parseResponse(JSON_REPORT_THERMAL_BEARING_MOCK);
    });

    it('should extract temperatures from thermal bearing response', () => {
      expect(result.temperatures).toEqual([
        {
          designation: 'Heating temperature',
          abbreviation: 'T',
          value: '58.7',
          unit: '°C',
        },
      ]);
    });

    it('should include temperatures in the result', () => {
      expect(result.temperatures).toBeDefined();
      expect(result.temperatures).toHaveLength(1);
    });
  });
});
