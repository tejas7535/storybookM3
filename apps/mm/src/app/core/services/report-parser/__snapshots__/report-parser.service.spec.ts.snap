// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReportParserService parse Response with checknut and clearances when result is defined should have empty notes section 1`] = `Array []`;

exports[`ReportParserService parse Response with checknut and clearances when result is defined should parse clearance classes data 1`] = `
Array [
  Object {
    "designation": "Check values for clearance classes ( CN )",
    "isImportant": true,
    "unit": "µm",
    "value": "300,0",
  },
  Object {
    "designation": "Check values for clearance classes ( C3 )",
    "isImportant": true,
    "unit": "µm",
    "value": "430,0",
  },
  Object {
    "designation": "Check values for clearance classes ( C4 )",
    "isImportant": true,
    "unit": "µm",
    "value": "640,0",
  },
]
`;

exports[`ReportParserService parse Response with checknut and clearances when result is defined should parse locknut data 1`] = `
Array [
  Object {
    "abbreviation": undefined,
    "designation": "Locknut",
    "unit": undefined,
    "value": "HM30/1000-H",
  },
  Object {
    "abbreviation": "D_m",
    "designation": "Outer diameter",
    "unit": "mm",
    "value": "1140,0",
  },
  Object {
    "abbreviation": "d_2",
    "designation": "Inner diameter",
    "unit": "mm",
    "value": "1000,0",
  },
  Object {
    "abbreviation": "b_1",
    "designation": "Width",
    "unit": "mm",
    "value": "100,0",
  },
  Object {
    "abbreviation": "m",
    "designation": "Groove width",
    "unit": "mm",
    "value": "60,0",
  },
  Object {
    "abbreviation": "n",
    "designation": "Groove depth",
    "unit": "mm",
    "value": "25,0",
  },
  Object {
    "abbreviation": undefined,
    "designation": "Thread",
    "unit": undefined,
    "value": "Tr1000x8",
  },
  Object {
    "abbreviation": undefined,
    "designation": "Lock washer",
    "unit": undefined,
    "value": "MS30/1000",
  },
]
`;

exports[`ReportParserService parse Response with checknut and clearances when result is defined should parse radial clearance data 1`] = `
Array [
  Object {
    "designation": "Axial displacement ( Min )",
    "isImportant": true,
    "unit": "µm",
    "value": "6300,0",
  },
  Object {
    "designation": "Axial displacement ( Max )",
    "isImportant": true,
    "unit": "µm",
    "value": "8500,0",
  },
  Object {
    "designation": "Radial clearance reduction by mounting ( Min )",
    "isImportant": true,
    "unit": "µm",
    "value": "410,0",
  },
  Object {
    "designation": "Radial clearance reduction by mounting ( Max )",
    "isImportant": true,
    "unit": "µm",
    "value": "550,0",
  },
]
`;

exports[`ReportParserService parse Response with checknut and clearances when result is defined should parse sleeve connectors data 1`] = `
Array [
  Object {
    "abbreviation": undefined,
    "designation": "Sleeve connector",
    "unit": undefined,
    "value": "PUMP.SLEEVE-CONNECTOR-G1/8",
  },
  Object {
    "abbreviation": undefined,
    "designation": "Number of sleeve connectors",
    "unit": undefined,
    "value": "1",
  },
]
`;

exports[`ReportParserService parseResponse when result is defined should have empty errors 1`] = `Array []`;

exports[`ReportParserService parseResponse when result is defined should have inputs 1`] = `
Array [
  Object {
    "hasNestedStructure": false,
    "subItems": Array [
      Object {
        "abbreviation": undefined,
        "designation": "Designation",
        "hasNestedStructure": false,
        "title": undefined,
        "titleID": undefined,
        "unit": undefined,
        "value": "2205-K-2RS-TVH-C3",
      },
      Object {
        "abbreviation": undefined,
        "designation": "Design",
        "hasNestedStructure": false,
        "title": undefined,
        "titleID": undefined,
        "unit": undefined,
        "value": "Self-aligning ball bearing",
      },
      Object {
        "abbreviation": undefined,
        "designation": "Series",
        "hasNestedStructure": false,
        "title": undefined,
        "titleID": undefined,
        "unit": undefined,
        "value": "22",
      },
      Object {
        "abbreviation": "B",
        "designation": "Width",
        "hasNestedStructure": false,
        "title": undefined,
        "titleID": undefined,
        "unit": "mm",
        "value": "18,000",
      },
      Object {
        "abbreviation": "D",
        "designation": "Outside diameter",
        "hasNestedStructure": false,
        "title": undefined,
        "titleID": undefined,
        "unit": "mm",
        "value": "52,000",
      },
      Object {
        "abbreviation": "d",
        "designation": "Inside diameter",
        "hasNestedStructure": false,
        "title": undefined,
        "titleID": undefined,
        "unit": "mm",
        "value": "25,000",
      },
      Object {
        "abbreviation": undefined,
        "designation": "Bore type",
        "hasNestedStructure": false,
        "title": undefined,
        "titleID": undefined,
        "unit": undefined,
        "value": "conical 1:12",
      },
    ],
    "title": "Bearing data",
    "titleID": "STRING_OUTP_BEARING_DATA",
  },
  Object {
    "hasNestedStructure": false,
    "subItems": Array [
      Object {
        "abbreviation": "d_sft",
        "designation": "Inner shaft diameter",
        "hasNestedStructure": false,
        "title": undefined,
        "titleID": undefined,
        "unit": "mm",
        "value": "0,000",
      },
      Object {
        "abbreviation": undefined,
        "designation": "Shaft material",
        "hasNestedStructure": false,
        "title": undefined,
        "titleID": undefined,
        "unit": undefined,
        "value": "steel",
      },
      Object {
        "abbreviation": "E",
        "designation": "Modulus of Elasticity",
        "hasNestedStructure": false,
        "title": undefined,
        "titleID": undefined,
        "unit": "N/mm²",
        "value": "210000",
      },
      Object {
        "abbreviation": "Nue",
        "designation": "Poisson ratio",
        "hasNestedStructure": false,
        "title": undefined,
        "titleID": undefined,
        "unit": undefined,
        "value": "0,30",
      },
    ],
    "title": "Shaft data",
    "titleID": "STRING_OUTP_SHAFT_DATA",
  },
  Object {
    "hasNestedStructure": false,
    "subItems": Array [
      Object {
        "abbreviation": undefined,
        "designation": "Bearing seat",
        "hasNestedStructure": false,
        "title": undefined,
        "titleID": undefined,
        "unit": undefined,
        "value": "tapered shaft",
      },
      Object {
        "abbreviation": undefined,
        "designation": "Measuring method",
        "hasNestedStructure": false,
        "title": undefined,
        "titleID": undefined,
        "unit": undefined,
        "value": "radial clearance reduction",
      },
      Object {
        "abbreviation": undefined,
        "designation": "Mounting method",
        "hasNestedStructure": false,
        "title": undefined,
        "titleID": undefined,
        "unit": undefined,
        "value": "hydraulically supported with hydraulic nut",
      },
    ],
    "title": "Mounting",
    "titleID": "STRING_OUTP_MOUNTING",
  },
]
`;

exports[`ReportParserService parseResponse when result is defined should have notes 1`] = `
Array [
  "The calculated values only apply for steel shafts that are either solid or have a bore diameter not bigger than the half outer diameter.",
]
`;

exports[`ReportParserService parseResponse when result is defined should have nut data 1`] = `
Array [
  Object {
    "abbreviation": undefined,
    "designation": "Hydraulic nut type",
    "unit": undefined,
    "value": "not available",
  },
  Object {
    "abbreviation": "A_k",
    "designation": "Piston face",
    "unit": "cm²",
    "value": "0,000",
  },
]
`;

exports[`ReportParserService parseResponse when result is defined should have pumps with recommended and alternative boolean flag 1`] = `
Object {
  "items": Array [
    Object {
      "field": "Pump for hydraulic nut",
      "isRecommended": true,
      "value": "PUMP1000-2,2",
    },
    Object {
      "field": "Pump for hydraulic nut",
      "isRecommended": false,
      "value": "PUMP700-2L",
    },
    Object {
      "field": "Pump for hydraulic nut",
      "isRecommended": false,
      "value": "PUMP1000-5L-AIR",
    },
  ],
  "title": "Pump for hydraulic nut",
}
`;

exports[`ReportParserService parseResponse when result is defined should have warnings 1`] = `
Array [
  "There is no hydraulic nut available for the selected bearing.",
  "  · 2205-K-2RS-TVH-C3",
]
`;

exports[`ReportParserService parseResponse when result is defined should parse mounting recommendations data 1`] = `
Array [
  "The measurement of radial clearance with a feeler gage is hardly possible for radial self-aligning ball bearings.",
  "Radial self-aligning ball bearings with an inner diameter up to 75 mm are mounted with a special tool (LOCKNUT-DOUBLEHOOK).",
  "The desired radial clearance results from the given torque or the torsion angle.",
  "The detailed mounting procedure is described in the manual BA28.",
  "some nested text",
]
`;

exports[`ReportParserService parseResponse when result is defined should parse response end positions data 1`] = `
Array [
  Object {
    "abbreviation": "dx_mnt",
    "designation": "Displacement (start to end position)",
    "isImportant": true,
    "unit": "mm",
    "value": "0,190",
  },
  Object {
    "abbreviation": "delta_sr",
    "designation": "Radial clearance reduction",
    "isImportant": false,
    "unit": "µm",
    "value": "8,7",
  },
]
`;

exports[`ReportParserService when parsing input with all possible identifier types should parse inputs 1`] = `
Array [
  Object {
    "hasNestedStructure": true,
    "subItems": Array [
      Object {
        "hasNestedStructure": true,
        "subItems": Array [
          Object {
            "designation": "apple",
            "hasNestedStructure": false,
            "unit": "kg",
            "value": "delicious apple",
          },
          Object {
            "designation": "bannana",
            "hasNestedStructure": false,
            "unit": "qt",
            "value": "delicious bannana",
          },
          Object {
            "designation": "tomato",
            "hasNestedStructure": false,
            "unit": undefined,
            "value": "some cherry tomatoes",
          },
          Object {
            "designation": "resultValues.designation",
            "hasNestedStructure": false,
            "unit": undefined,
            "value": "designation",
          },
        ],
        "title": undefined,
        "titleID": undefined,
      },
    ],
    "title": "inner Block",
    "titleID": undefined,
  },
  Object {
    "hasNestedStructure": false,
    "subItems": Array [
      Object {
        "abbreviation": undefined,
        "designation": "reference rating life",
        "hasNestedStructure": false,
        "title": undefined,
        "titleID": undefined,
        "unit": undefined,
        "value": "nominal",
      },
    ],
    "title": "title variable Block 1",
    "titleID": undefined,
  },
  Object {
    "hasNestedStructure": false,
    "subItems": Array [
      Object {
        "abbreviation": "BPFFO",
        "designation": "Overrolling frequency factor on outer ring",
        "hasNestedStructure": false,
        "title": undefined,
        "titleID": undefined,
        "unit": undefined,
        "value": "4.1181",
      },
      Object {
        "abbreviation": "BPFFI",
        "designation": "Overrolling frequency factor on inner ring",
        "hasNestedStructure": false,
        "title": undefined,
        "titleID": undefined,
        "unit": undefined,
        "value": "5.8819",
      },
      Object {
        "abbreviation": "BSFF",
        "designation": "Overrolling frequency factor on rolling element",
        "hasNestedStructure": false,
        "title": undefined,
        "titleID": undefined,
        "unit": undefined,
        "value": "2.7465",
      },
    ],
    "title": "Basic frequencies factor",
    "titleID": "STRING_OUTP_BASIC_FREQUENCIES_FACTOR",
  },
]
`;
